### PR TITLE
Add request validation to chs-delta-api

### PR DESCRIPTION
**Pull request by amoriartyCH**
**Add validation using kin-openAPI3 lib to chs-delta-api -- Resolves DSND-87**

Summary:
This pull request adds the use of the approved lib [kin-openAPI3](https://github.com/getkin/kin-openapi) to the chs-delta-api. The library takes a http request and uses a provided open API 3 spec to validate it. If errors are found in the request then an array of [Companies House error objects](https://developer-specs.company-information.service.gov.uk/companies-house-public-data-api/resources/error?v=latest)  is returned to the requester detailing what each error is.

Details of pull request:
- Add new validation package to house all validation logic.
- Minor changes to api-spec to account for token name change and enum values needing to be wrapped in quotes.
- Changes to OfficerDeltaHandler to account for new call to validate request and handle all possible responses.
- Add struct for new CH Error object used in validation.
- Improve logging and comments for entire project.
- Add unit tests where needed and improve coverage where possible.
- Bump version of application.

- Bump Go version to 1.16, we are using this service as the initial trail to roll out Go 1.16 as it's a small service.

### CODE REVIEW

File: validation/chOpenApiValidator.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+// NewCHValidator returns a new CHValidator implementation.
+func NewCHValidator() CHValidator {
+	return CHValidatorImpl{}
+}
+
+// ValidateRequestAgainstOpenApiSpec takes a request and an openAPI spec location (string relative path) and uses the
+// spec to validate the provided request. If any validation errors are found, then they are formatted and returned to the
+// caller. If any errors are encountered while attempting to validate, they are handled and also returned to the caller.
+func (chv CHValidatorImpl) ValidateRequestAgainstOpenApiSpec(httpReq *http.Request, openApiSpec string) ([]byte, error) {
```
Comment (emead) [inline validation/chOpenApiValidator.go:43]: Would it be possible for this validation package to be stored in the chs.go repository to enable re-use by other services?

File: handlers/officerDeltaHandler_test.go [hunk 2]
```diff
@@ -139,17 +171,116 @@ func TestOfficerDeltaHandlerFailsSend(t *testing.T) {
...
+			Convey("Then the response should be 500 and an error returned", func() {
+				So(resp.Code, ShouldEqual, http.StatusInternalServerError)
+			})
+
+			destroyEnv()
+		})
+	})
+}
+
+// TestOfficerDeltaHandlerErrorsCallingValidation asserts that the officerDeltaHandler returns an internal error status when validation fails
```
Comment (divya051988) [inline handlers/officerDeltaHandler_test.go:199]: If the validation fails then it should be a bad request right ?

File: handlers/officerDeltaHandler_test.go [hunk 3]
```diff
@@ -139,17 +171,116 @@ func TestOfficerDeltaHandlerFailsSend(t *testing.T) {
...
+			cfg, _ := config.Get()
+
+			handler := NewOfficerDeltaHandler(svc, h, chv, cfg)
+
+			errBytes := []byte("error string")
+			chv.EXPECT().ValidateRequestAgainstOpenApiSpec(req, handler.cfg.OpenApiSpec).Return(errBytes, nil)
+
+			handler.ServeHTTP(resp, req)
+
+			Convey("Then the response should be 500 and an error returned", func() {
```
Comment (divya051988) [inline handlers/officerDeltaHandler_test.go:267]: This should say 400

File: services/kafkaService.go [hunk 4]
```diff
@@ -95,7 +95,7 @@ func (kSvc *KafkaServiceImpl) SendMessage(topic, data string) error {
 
 	// Construct a chs-delta using provided data.
 	deltaData := models.ChsDelta{
-		ContextId: "uu-aa-dd", // TODO: Create contextId.
+		ContextId: "uu-aa-dd", // TODO: Use ERIC-Request-Id header and change logs to use C versions. Also add Request-Id to response.
```
Comment (divya051988) [inline services/kafkaService.go:98]: Please remove TODO's from the code.

File: validation/chOpenApiValidator_test.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/getkin/kin-openapi/openapi3filter"
+	"github.com/getkin/kin-openapi/routers"
+	router "github.com/getkin/kin-openapi/routers/gorillamux"
+	. "github.com/smartystreets/goconvey/convey"
+	"net/http/httptest"
+	"testing"
+)
+
+const (
```
Comment (divya051988) [inline validation/chOpenApiValidator_test.go:16]: Is it possible to load this from a file ?

File: validation/chOpenApiValidator_test.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+	requestBody = `{
+  "officers": [{
+    "company_number": "09876543",
```
Comment (divya051988) [inline validation/chOpenApiValidator_test.go:19]: Can we use a fake company 00000000 or something ?

File: handlers/officerDeltaHandler_test.go (hunk 2 above)
Comment (amoriartyCH) [reply handlers/officerDeltaHandler_test.go]: Comment was updated to clarify that it is indeed when call to validate fails, not a user error.

File: handlers/officerDeltaHandler_test.go (hunk 3 above)
Comment (amoriartyCH) [reply handlers/officerDeltaHandler_test.go]: Updated and resolved.

File: services/kafkaService.go (hunk 4 above)
Comment (amoriartyCH) [reply services/kafkaService.go]: Removed.

File: validation/chOpenApiValidator_test.go (hunk 5 above)
Comment (amoriartyCH) [reply validation/chOpenApiValidator_test.go:16]: Changed to just be a small dummy data request.

**Review by divya051988 — APPROVED** [review]

**Pull request merged**